### PR TITLE
sys/app_metadata: Add a way to print application metadata

### PR DESCRIPTION
**Pull request by MrKevinWeiss**
### Contribution description
This PR adds a command to print a number of useful application parameters such as BOARD, MCU, OS_VERSION, APP_NAME, etc.
The print function prints in json as it is human readable (enough) and machine readable/easy to format.
It also adds an optional parameter APP_SHELL_FMT (application shell format) intended to help coordinate any specific formatting that the application is following.  It can be specified in the makefile or in commandline but if not specified it doesn't print anything.

### Testing procedure
run `BOARD=<???> make flash term -C tests/shell/` or `APP_SHELL_FMT="test" BOARD=<???> make flash term -C tests/shell/` or add APP_SHELL_FMT="test" in the makefile.
Then in the shell check the response when calling `app_metadata`
Or just call `make flash test -C tests/shell/`

### Issues/PRs references
None

### CODE REVIEW

**Comment by miri64** [comment]
Meta data on what?

**Comment by miri64** [comment]
> Meta data on what?

After looking at the code: how about renaming the module `app_metadata` to make this clearer.

File: sys/Makefile.include [hunk 1]
```diff
@@ -31,6 +31,13 @@ ifneq (,$(filter oneway_malloc,$(USEMODULE)))
   USEMODULE_INCLUDES += $(RIOTBASE)/sys/oneway-malloc/include
 endif
 
+ifneq (,$(filter metadata,$(USEMODULE)))
+  # Overwrite the interface version
```
Comment (kaspar030) [inline sys/Makefile.include]: what interface is this referring to? API?

File: sys/include/metadata.h [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+#define METADATA_H
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief Prints information about metadata to stdio.
+ */
+void metadata_get(void);
```
Comment (kaspar030) [inline sys/include/metadata.h]: "get" implies this goes somewhere. maybe "print" would be more appropriate?

File: sys/Makefile.include (hunk 1 above)
Comment (MrKevinWeiss) [reply sys/Makefile.include]: None right now, the plan it to have the RDM merged then we can specify it adheres to the rules.  Similar to something like it adheres to a [JSON schema draft 7](https://json-schema.org/draft-07/json-schema-release-notes.html).

File: sys/include/metadata.h (hunk 2 above)
Comment (MrKevinWeiss) [reply sys/include/metadata.h]: Yup, that what I had originally but saw something with get... I will change it!

File: sys/Makefile.include (hunk 1 above)
Comment (miri64) [reply sys/Makefile.include]: What RDM? There are 3 under review right now?

**Comment by MrKevinWeiss** [comment]
@miri64 sure thing, I wasn't that happy with the naming anyways. app_metadata it will be.

File: sys/Makefile.include (hunk 1 above)
Comment (miri64) [reply sys/Makefile.include]: > None right now, the plan it to have the RDM merged then we can specify it adheres to the rules. Similar to something like it adheres to a [JSON schema draft 7](https://json-schema.org/draft-07/json-schema-release-notes.html).

Though I like the simplicity of this implementation, I wonder

a) Why pin-point to JSON, when we also support e.g. CBOR?
b) Why use `printf` and not `fmt` *iff* we want to stick to JSON.

**Comment by kaspar030** [comment]
I must ask for the use of this. Having the "metadata" printed is quite useless unless for the shell.

How about containing this in sys/shell/commands?

I'd name it "about". Everyone seeing a prompt will type "help", and everyone will try "about".

File: sys/Makefile.include (hunk 1 above)
Comment (MrKevinWeiss) [reply sys/Makefile.include]: The RDM I am referring to is #10624

**Comment by MrKevinWeiss** [comment]
@kaspar030 Ok, is "about" fine with you @miri64

**Comment by miri64** [comment]
> @kaspar030 Ok, is "about" fine with you @miri64

Don't care much about that :-)

**Comment by miri64** [comment]
> Don't care much about that :-)

I mean the name of the shell command. The module should still be named `app_metadata` or something like that.

**Comment by miri64** [comment]
`about` as a module name would be too unspecific.

**Comment by kaspar030** [comment]
> `about` as a module name would be too unspecific.

that's why I suggest to just put the code into ```sys/shell/commands/sc_about.c```. Not even a module needed.

**Comment by miri64** [comment]
> that's why I suggest to just put the code into `sys/shell/commands/sc_about.c`. Not even a module needed.

I prefer to always have a corresponding module to a shell command. This way we have a way to later integrate it into other configuration/information mechanisms (e.g. via CoAP).

**Comment by MichelRottleuthner** [comment]
Actually I would have done it like @kaspar030 wrote above, i.e. just adding a `sc_about.c`. But thinking about it, I agree that putting it to a separate module is more flexible. 
Another thing: The dedicated "test" can probably be removed as it doesn't really add something. `sys/shell` should be enough, no?

**Comment by MrKevinWeiss** [comment]
Also isn't the USEMODULE the way to select which shell commands are used (I tried to model this after the ps command)

**Comment by kaspar030** [comment]
I just realized that the module outputs in json format...

Look, my first reaction was as @miri64's: "what metadata?" 

I know how this works. You're happily solving some problem (in this case, maybe you've been working on the unified testing RDM) and need some functionality. It has something to do with metadata, and 15min later there's a PR for "sys/metadata".

This unfortunately blanks out that there's a plethora of different understandings for metadata. The PR description is not very helpful (it could mention json format, and that it is supposed to be used by the tests, ...). So as is, there's "metadata" in "sys/", but noone gets what the folder is there for.

I suspect there are gonna be more related functions. So how about directly creating a "sys/test" module? IMO, it is the right place for a function to print test metadata. It could be the in-tree code support for the coming RDM.
Soon there'll be "test_print_result_value()", "test_print_success()", ... Surely we don't want "sys/result_value" as module.

**Comment by miri64** [comment]
> Also isn't the USEMODULE the way to select which shell commands are used (I tried to model this after the ps command)

Yepp, but in theory you could just define a pseudo-module for that (e.g. like I did to provide the `pktbuf` command -- which is still implemented in `gnrc_pktbuf_static`, but usually not exposed to the shell per default).

**Comment by MrKevinWeiss** [comment]
@kaspar030 Heh, correct you got me :grin: and I will improve the description of the PR.  I added this because I thought it would be useful regardless to have this from an application perspective.  Yes the formatting is json but it is also somewhat human readable which has been the standard for most of this.

@MichelRottleuthner the test was more to show how to use it.  It can also be used just for testing and dropped before merge (I think @cladmi is doing something like that).  As you wish.

In terms of what is metadata, you see the information that I am wanting.  Just a mechanism for the application to identify itself, I am really not tied to the name and appreciate anyone input (or simply say do it like this)

I can also move it to sys/tests but I am more thinking we should have some consensus.  For me it seems like there is more than one way to skin a cat but maybe we can hit two birds with one stone... :)

@miri64 Thanks for the pseudo-module hint!

**Comment by miri64** [comment]
> In terms of what is metadata, you see the information that I am wanting. Just a mechanism for the application to identify itself, I am really not tied to the name and appreciate anyone input (or simply say do it like this)

Yes, we can see that *after* looking at the code. But the name of the module, the description of the shell command, and the descriptive text of the PR should also make that clear, so people don't assume "Oh I need metadata for feature X, and there is a metadata module. It is probably supposed to go there?!??"

**Comment by MrKevinWeiss** [comment]
@miri64 Yup, bad name, got it.  Any suggestions?  The "about" command also seems reasonable.  Maybe "app_info"?

**Comment by miri64** [comment]
> @miri64 Yup, bad name, got it. Any suggestions? The "about" command also seems reasonable. Maybe "app_info"?

`app_metadata` is fine with me. Just `metadata` was a little bit confusing to me.

**Comment by MrKevinWeiss** [comment]
I am fine with that as well!

**Comment by miri64** [comment]
Also, please find a more specific name than `INTERFACE_VERSION` or interface. It's already hard enough for people to find out if others talk about network interfaces or APIs, so calling a specific API just "*the interface*" is very confusing.

**Comment by MrKevinWeiss** [comment]
Yup yup.  Just need to think of what is specific enough!

File: sys/Makefile.include [hunk 3]
```diff
@@ -31,6 +31,13 @@ ifneq (,$(filter oneway_malloc,$(USEMODULE)))
   USEMODULE_INCLUDES += $(RIOTBASE)/sys/oneway-malloc/include
 endif
 
+ifneq (,$(filter app_metadata,$(USEMODULE)))
+  # Overwrite the interface version
```
Comment (MichelRottleuthner) [inline sys/Makefile.include]: The comment `# Overwrite the interface version` is still using the very unspecific wording `interface version`. How about something more explicit about the fact that this modifies the reported version of the standardized shell output format? (and to explain the context maybe a reference to the RDM?)

**Comment by MrKevinWeiss** [comment]
:kissing: Thanks for the review @MichelRottleuthner

**Comment by miri64** [comment]
> @kaspar030 , @miri64 you agree?.

For reference: yepp, all comments were addressed :-)

**Pull request merged**